lib/src/solver/version_solver.dart

```diff
@@ -2,875 +2,417 @@
-    }
+    var result = await _result();
 
-    var required = _selection.getRequiredDependency(name);
-    if (required != null && !package.samePackage(required.dep)) return null;
+    // Gather some solving metrics.
+    log.solver('Version solving took ${stopwatch.elapsed} seconds.\n'
+        'Tried ${_solution.attemptedSolutions} solutions.');
 
-    return package;
+    return result;
   }
 
-  /// Tries to find the best set of versions that meet the constraints.
+  /// Performs [unit propagation][] on incompatibilities transitively related to
+  /// [package] to derive new assignments for [_solution].
   ///
-  /// Selects matching versions of unselected packages, or backtracks if there
-  /// are no such versions.
-  Future<List<PackageId>> _solve() async {
-    // TODO(nweiz): Use real while loops when issue 23394 is fixed.
-    await Future.doWhile(() async {
-      // Avoid starving the event queue by waiting for a timer-level event.
-      await new Future(() {});
-
-      // If there are no more packages to traverse, we've traversed the whole
-      // graph.
-      var ref = _selection.nextUnselected;
-      if (ref == null) return false;
-
-      var queue;
-      try {
-        queue = await _versionQueueFor(ref);
-      } on SolveFailure catch (error) {
-        // TODO(nweiz): adjust the priority of [ref] in the unselected queue
-        // since we now know it's problematic. We should reselect it as soon as
-        // we've selected a different version of one of its dependers.
-
-        // There are no valid versions of [ref] to select, so we have to
-        // backtrack and unselect some previously-selected packages.
-        if (await _backtrack()) return true;
-
-        // Backtracking failed, which means we're out of possible solutions.
-        // Throw the error that caused us to try backtracking.
-        if (error is! NoVersionException) rethrow;
-
-        // If we got a NoVersionException, convert it to a
-        // non-version-specific one so that it's clear that there aren't *any*
-        // acceptable versions that satisfy the constraint.
-        throw new NoVersionException(error.package, null,
-            (error as NoVersionException).constraint, error.dependencies);
+  /// [unit propagation]: https://github.com/dart-lang/pub/tree/master/doc/solver.md#unit-propagation
+  void _propagate(String package) {
+    var changed = new Set.from([package]);
+
+    while (!changed.isEmpty) {
+      var package = changed.first;
+      changed.remove(package);
+
+      // Iterate in reverse because conflict resolution tends to produce more
+      // general incompatibilities as time goes on. If we look at those first,
+      // we can derive stronger assignments sooner and more eagerly find
+      // conflicts.
+      for (var incompatibility in _incompatibilities[package].reversed) {
+        var result = _propagateIncompatibility(incompatibility);
+        if (result == #conflict) {
```

Comment:
A symbol literal?

I think this would be clearer if it returned some enum-ish class that contained the string when that case is matched. Also, document when it may return a string.

Reply:
> I think this would be clearer if it returned some enum-ish class that contained the string when that case is matched.

Making this type-safe is sooooo much boilerplate for one private function 😩. It takes *three separate classes* to define a type that has both enum members and members that have values. I'll do it if you really think it's important, but I'd much rather be terse if possible.

My kingdom for algebraic structs.

> Also, document when it may return a string.

Done.